extensions/pkg/terraformer/state.go

```diff
@@ -91,8 +98,30 @@ func (t *terraformer) GetStateOutputVariables(ctx context.Context, variables ...
 	return output, nil
 }
 
-// IsStateEmpty returns true if the Terraform state is empty, and false otherwise.
+// IsStateEmpty returns true if the Terraform state is empty and the terraformer finalizer
+// is not present on any of the used configmpas and secrets. Otherwise, it returns false.
 func (t *terraformer) IsStateEmpty(ctx context.Context) bool {
+	for _, obj := range []client.Object{
+		&corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Namespace: t.namespace, Name: t.configName}},
+		&corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Namespace: t.namespace, Name: t.stateName}},
+		&corev1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: t.namespace, Name: t.variablesName}},
+	} {
+		resourceName := obj.GetName()
+		if err := t.client.Get(ctx, kutil.Key(t.namespace, resourceName), obj); client.IgnoreNotFound(err) != nil {
+			t.logger.Error(err, "failed to get resource", "name", resourceName)
+			return false
+		}
+
+		resourceHasFinalizer, err := extensionscontroller.HasFinalizer(obj, TerraformerFinalizer)
```

Comment:
I see that currently we have multiple HasFinalizer funcs. One of them is deprecated in favour of `sigs.k8s.io/controller-runtime/pkg/controller/controllerutil.ContainsFinalizer` after the controller-runtime@v0.7 revendoring. Should we use directly `controllerutil.ContainsFinalizer`?

```suggestion
		resourceHasFinalizer := controllerutil.ContainsFinalizer(obj, TerraformerFinalizer)
```

Reply:
Also noticed this. We can do it directly for this case, but I anyway have a general cleanup of this on my list as part of https://github.com/gardener/gardener/issues/3109.

Reply:
I am having hard times when I have to chose between the multiple implementations of the same util functions. 
So I am open to this suggestion and have applied it.